### PR TITLE
Random space structures

### PR DESCRIPTION
**Pull request by unid15**
- At the beginning of each round, a random number of pre-mapped structures (later - vaults) is placed around the map
- Exact amount is rand(1,(number of possible vaults)), maximum amount is 10 (unless somebody adds more areas)
- Vaults are stored in maps/randomVaults/ as .dmm files
- The vaults are only placed in designated areas around the map
  
- Each area can only house one vault inside
- This PR includes 5 high quality vaults:  [stash in an asteroid (guns=toys)](http://puu.sh/nwZbx/0bff8a9aa9.png), [ice cream truck crash](http://puu.sh/nwZdG/f1712a4673.png), [hivebot factory](http://puu.sh/nAB8C/101072fa8f.png), [clown base](http://puu.sh/nBQ2I/6c98716711.png), [ruined RUST generator by @PJB3005](https://camo.githubusercontent.com/ad226e04acbfb21427a160c065b5b1bd81b50659/68747470733a2f2f646c2e64726f70626f7875736572636f6e74656e742e636f6d2f752f3130373032383032362f5368617265582f323031362f30332f647265616d6d616b65725f323031362d30332d31315f32312d34332d31352e706e67)
- Ice cream truck crash has 3 temp guns (guarded by 6 polar bears)
- Asteroid stash has some chest mimics and a bat helmet
- Hivebot factory has three superglue bottles, a fireaxe, an assault rifle and a syndicate flaregun. The assault rifle is guarded by tons of hivebots, superglue is guarded by two cyber horrors. Fireaxe and syndicate flaregun are hidden in a tiny r-wall room with a cyberhorror.
- Clown base has a bunch of powerful banana peels, a buttbot, H.O.N.K weapons and a genuine clown PDA hidden

**STATUS:**
- works and 100% tested
- vaults may still need some work

### CODE REVIEW

**Comment by Shadowmech88** [comment]
:+1:

**Comment by PJB3005** [comment]
OH SHIT SON :+1:

**Comment by ghost** [comment]
Aha! I'd this idea for #6789 and it's good this'll exist for that. Perhaps consider using datums for the sorting process, though? Otherwise, :+1: :ok_hand:

**Comment by Shadowmech88** [comment]
Hypothetically, how difficult would it be to procedurally generate the structures themselves? Or have a single vault that's the designated "random" vault, and just procedurally generate that specific one if the map spawns it?

**Comment by ghost** [comment]
@Shadowmech88 Not difficult if you want it to be randomly generated. Very difficult if you don't want that random generation to not look like complete ass

**Comment by Shadowmech88** [comment]
What about doing that thing that Spelunky does where you pre-make chunks, and then have the engine slap some of the chunks together in a random order?

**Comment by PJB3005** [comment]
Same issue @Shadowmech88 because this isn't Spelunky.

**Comment by ghost** [comment]
Is there any way we can make random "vaults" for this?

Got ideas for a few large ones, small ones, etc.

**Comment by icantthinkofanameritenow** [comment]
@Skullyton 
I think maybe it could work if we borrowed the stuff used for generating the hidden rooms on the mining asteroid, but I'm not entirely sure. 

I support this idea- it'll give people a good reason to go and explore space beyond looking for the clown roid.

**Comment by unid15** [comment]
Randomly generated vaults are gonna work on a different code than this, that's for sure! And it's gonna take a lot of work.
Anyways, I just _100%_ tested this and it seems to be working fine. Currently these are the 3 vaults that can spawn (spoiler alert they're shit):

[doomed soviet satelite](http://puu.sh/nwZ8O/957da2bf2c.png), [stash in an asteroid (guns=toys)](http://puu.sh/nwZbx/0bff8a9aa9.png), [ice cream truck crash](http://puu.sh/nwZdG/f1712a4673.png)

**Comment by 9600bauds** [comment]
Dude this is sweet as FUCK, but please for the love of God spread out the spawners over the many z-levels we have

**Comment by ghost** [comment]
@unid15 Nice, how do we make those tho?

**Comment by unid15** [comment]
They're already spread out over all 5 z-levels

**Comment by 9600bauds** [comment]
Oh alright, I saw the screenshot of all of them spawning in the same z-level and couldn't gleam from the code if it was that way or not

**Comment by unid15** [comment]
@Skullyton they're .dmm map files, so they can be made in Dream Maker (just like normal maps). Just don't make them too big

**Comment by icantthinkofanameritenow** [comment]
@unid15 

They're decent proofs of concept at least. The full ones should probably have some more appealing loot, though. Maybe the relatively weaker adminbus items that more or less function like recolors of existing gear, like the T-51b power armor.

**Comment by CptWad** [comment]
SWEET
How easy is it to add more possible things to the list of possible spawners?

**Comment by unid15** [comment]
You just have to put the file into a folder and add its filename to the list in code/modules/randomMaps/vaults.dm

**Comment by CptWad** [comment]
noice

**Comment by clusterfack** [comment]
Does this move the station, or is the station usually not centered? Because its kind of triggering me.

**Comment by CptWad** [comment]
Thats derelict and its been like that since the space quadrupling

**Comment by CptWad** [comment]
Wheres restraunt.dmm?

**Comment by SkowronX** [comment]
Derelict plasmaman vessels with hostile plasmamen when

**Comment by Exxion** [comment]
Holy SHIT

**Comment by FudgePucker** [comment]
The ice cream truck needs some better loot if its gonna be guarded by six bears

**Comment by CptWad** [comment]
How bout a temp gun? for the ice cream truck?

**Comment by Probe1** [comment]
:+1: 

@clusterfack the station was always REAL CLOSE to the left side of the map.  I've had trouble fitting big custom made ships at the ferry dock because of the z level border.

File: code/controllers/master_controller.dm [hunk 1]
```diff
@@ -225,6 +225,15 @@ datum/controller/game_controller/proc/cachedamageicons()
 	else
 		log_startup_progress("Not generating minimaps - SKIP_MINIMAP_GENERATION found in config/config.txt")
 
+	if(!config.skip_vault_generation)
+		spawn()
```
Comment (PJB3005) [inline code/controllers/master_controller.dm]: Due to this `spawn()` the `Finished initializations` message will come up before this even starts to generate the structures.

File: code/modules/randomMaps/vaults.dm [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+//Because areas are shit and it's easier that way!
+
+//Each of these areas can only create ONE vault. Only using /area/random_vault/v1 for the entire map will result in ONE vault being created.
+//Placing them over (or even near) shuttle docking ports will sometimes result in a vault spawning on top of a shuttle docking port. This isn't a big problem, since
+//shuttles can destroy the vaults, but it's better to avoid that
+//If you want more vaults, feel free to add more subtypes of /area/random_vault. You don't have to add these subtypes to any lists or anything - just map it and the game will handle the rest.
+
+//"/area/random_vault" DOESN'T spawn any vaults!!!
+/area/random_vault/v1
```
Comment (PJB3005) [inline code/modules/randomMaps/vaults.dm:38]: Fucking areas.

THANKS BYOND.

To be honest this is probably the rare exception where I'm fine with map instances, but whatever.

**Comment by zashkuar** [comment]
roid stash is fine unlike others
Also, look into leaked goon code, they have some nice random stuff, like small abadoned factory or ship with servers and robots

**Comment by Shadowmech88** [comment]
Should we move the station to the center of its Z-level now that the Z-level's been expanded?

**Comment by unid15** [comment]
What are you guys talking about  it's in the middle

**Comment by icantthinkofanameritenow** [comment]
@CptWad

Have you already forgotten that temp guns are useless after Celt's hypothermia update? Even when emagged they're only good for killing slimes. Not to mention that something that can be mass produced with protolathes hardly makes up for fighting through six bears to get to it. 

Remember, the riskier it is to get to the loot the more rewarding it ought to be.

**Comment by 9600bauds** [comment]
A pre-emagged temp gun then

**Comment by icantthinkofanameritenow** [comment]
@zashkuar

Not really. It's just a Batman helmet (not even the full suit, at that), and a syndicate balloon along with some space cash. That's some lame loot if you ask me. 

If anything the satellite is a better example, since it has access to items impossible to find anywhere else outside of special circumstances (the crossbow and the Soviet vending machine).

**Comment by 9600bauds** [comment]
I'm going to stress this once again:
Randomized loot spawners are easy to implement and it would be a very good time to start using them more

**Comment by icantthinkofanameritenow** [comment]
@9600bauds
Even when emagged the temp gun is little more than a joke. It would be better to just replace it with something else that would justify the difficulty of getting to it. Either that, or remove the bears entirely since the rewards for killing them simply aren't worth it now. After all, you need to provide motivation to look for these vaults long after the thrill of exploration and discovery wears off- and the only way to do that is via the lure of loot.

Now that I think of it, what about all the guns that Deity added a while back that were never available outside of Summon Guns and adminbus? I'm sure a Mosin-Nagant wouldn't be out of place on the Soviet satellite. 

While it's on my mind, here's some relatively usable stuff I found in the code that could be used as rewards:

Advanced Clown Shoes
The Doomguy clothing (you'll want to add descriptions for them though, they have none as yet)
Ancient Space Suit
T-51b power armor
The load bearing equipment accessories (e.g. webbing)
The Adam Jensen costume for HoS
Tactical Turtleneck
Judge Armor
X-ray laser gun (it already exists as a xenoarch find)
Bluespace Crystals
Mind Batterer (in traitordevices.dm)
Switchtool (the standard one, not the surgeon's one)
Swiss Army Knife
Glow sticks
Batman belt

And I'm sure there's more we can think of too. 

> I'm going to stress this once again:
> Randomized loot spawners are easy to implement and it would be a very good time to start using them more.

Agreed. That said, I've looked at what several of them spawn (specifically the ones in Deff's maintainance) and I can't help but notice that they spawn trash or items much more easily obtained from other sources more often than not. Vault spawners should be more focused on generating rare items that would justify the difficulty of finding them and killing whatever might be guarding them.

**Comment by ghost** [comment]
> Have you already forgotten that temp guns are useless after Celt's hypothermia update? 

Please stop parroting this myth. They're no more or less effective than they were pre hypothermia. That said, they'd be a shite loot item because they are awful anyway.

**Comment by Arthurdentist** [comment]
instead of arguing about temp guns lets try and map some new space content to go with this. Here is my shit attempt which I threw together 10 minutes after reading this PR. 

devils bar

Its a good PR though and I hope it leads to a more interesting space experience™

**Comment by SysadminJeroen** [comment]
What about an abandoned space arcade with a very large chance (or 100%) for an Outbomb Cuban Pete machine?

**Comment by CptWad** [comment]
shit I was gonna do a few once it was merged so it was less of a hassle but I could whip some shit up
also whats the size limit for the things, not that ill push it but I just need to know

**Comment by 9600bauds** [comment]
Make the dice have a 10% chance to be an E20

**Comment by clusterfack** [comment]
Ho ho thats devious

**Comment by SysadminJeroen** [comment]
@9600bauds Maybe a derelict library?

**Comment by ihadtoregisterforthis** [comment]
How does this affect meteors?

**Comment by icantthinkofanameritenow** [comment]
@9600bauds 
Or better yet, the cursed dice.

**Comment by CptWad** [comment]
@ihadtoregisterforthis
if its on the station z-level it would sorta but on other z-levels it wont do shit

**Comment by 9600bauds** [comment]
A few problems I've found so far
- When the game starts, it says "skeleton" and a picture of a newcop. I don't know what that's supposed to be.
- Vaults can spawn in the station's z-level, is this intended?
- Well... The vaults show up in the minimaps if you clear your cache and login again.
  dreamseeker_2016-03-08_14-54-44

**Comment by unid15** [comment]
Yeah spawning on z-level 1 is intended. Will fix the other two (thanks for testing)

**Comment by unid15** [comment]
> When the game starts, it says "skeleton" and a picture of a newcop. I don't know what that's supposed to be.

oh god this makes **zero** sense

when you become an observer (only from the round-start screen, _BEFORE the round starts_), an invisible skeleton spawns in the fucking nullspace and shits a runtime whenever ANY sound is played

but if you spawn normally as a human, everything works perfectly and :100:% flawlessly

send help

**Comment by Shadowmech88** [comment]
> when you become an observer (only from the round-start screen), an invisible skeleton spawns in the fucking nullspace and shits a runtime whenever ANY sound is played

Literally what the fuck.

**Comment by unid15** [comment]
Literally this



 - runtime in the code above that spawns the skeleton

 - whenever any sound is played

 - even the beach

**Comment by PJB3005** [comment]
What the fuck? 

If that isn't a BYOND issue I don't know what is....

**Comment by FudgePucker** [comment]
'Spookie skeleton in nullspace' should be an arcade game that spawns only in a vault. Its attack will be runtimes.

**Comment by unid15** [comment]
I deleted the doomed satelite from my hard drive, removed all references to it in the code, sprayed my PC with holy water and suddenly everything started working again

If you guys have it saved on your computer for some reason, delete it now before it kills your family and turns you into a nullspace skeleton

(the problem was really in that map, I spawned 10 ice cream trucks and everything was fine, but 1 doomed satelite ruined everything)

**Comment by CptWad** [comment]
jesus christ how terrifying
Was there something on the satellite that caused all this or is it just a mystery

**Comment by Zth--** [comment]
Get a copy so we can make a creepypasta

**Comment by unid15** [comment]
Fixed minimaps for good, fixed lighting, added abandoned hivebot factory with good rewards (assault rifle, 3x superglue, fireaxe, flare gun) - the conveyor belts don't work



The doomed satelite still exists in the vaults folder

**Comment by 9600bauds** [comment]
How are those emitters powered

**Comment by unid15** [comment]
Vaults are always powered, like shuttles, so the answer is _magic_

**Comment by CptWad** [comment]
The Hivebot on the outside is going to wander off and end up on the station 50% of the time

**Comment by icantthinkofanameritenow** [comment]
Maybe replace the assault rifle with something more exotic? In an average round RnD makes a few dozen assault rifles at the bare minimum. What about a silenced pistol or something instead?

**Comment by Arthurdentist** [comment]
Could you generate ores in the 'asteroid' part of that vault with good chances for decent ores? Diamond et al?

**Comment by ghost** [comment]
That's a specific asteroid wall type, 'Good yield'.

The metaclub asteroid's made of the stuff, hence why you rarely see gibtonite on it

**Comment by ihadtoregisterforthis** [comment]
Would it be possible to have categories of structures so a very dangerous room doesn't threaten the non-explorers?

**Comment by jernaumorat** [comment]
I think we're going to need to expand the spacepod bay, shit's gonna get busy in there after this.

**Comment by ghost** [comment]
Spacepods would need to be better than a glorified jetpack first.

Main two things added to them that would make them perfect would be a contained storage for stashing loot, and some way of locking it beyond taking the cell out when you go wandering

**Comment by PJB3005** [comment]
Space pods you can walk inside when?

**Comment by unid15** [comment]
new stuff

http://puu.sh/nBQ2I/6c98716711.png

Banana peels are superglued, anchored and extra slippy, disposals bin throws you into space

Rewards are honkmech weapons, magivend, buttbot and a hidden clown PDA (plus the peels)

**Comment by Shadowmech88** [comment]
> the peels are a reward

But how would you take them if they're anchored?

**Comment by unid15** [comment]
You can pick up anchored items, you just can't pull them out of the way or put them in lockers

File: code/modules/randomMaps/vaults.dm [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			vault_map_names.Remove(map_name)
+
+			var/path_file = "[vault_map_directory][pick(map_name)].dmm"
+
+			if(fexists(path_file))
+				var/list/L = maploader.load_map(file(path_file), vault_z, vault_x, vault_y)
+				for(var/turf/new_turf in L)
+					new_turf.flags |= NO_MINIMAP //f u c k minimaps
+
+				message_admins("<span class='info'>Loaded [path_file]: [formatJumpTo(locate(vault_x, vault_y, vault_z))]!!!")
```
Comment (PJB3005) [inline code/modules/randomMaps/vaults.dm]: Don't you think 3 exclamation marks is a tad redundant?

**Comment by Shadowmech88** [comment]
> you can pick up anchored items

What property of projectiles prevents them from being picked up, then? Because they can't be picked up. I thought it was the anchoring that did it.

**Comment by unid15** [comment]
@Shadowmech88 they can't be clicked with the mouse[!](nothing here he-he)

**Comment by Shadowmech88** [comment]
@unid15 
Is there some mechanic that prevents you from clicking them?
What if you were to right-click and select "pick up" from there? Could you pick up a projectile frozen by time-stop, theoretically?

**Comment by Exxion** [comment]
I believe you can just make `attack_hand()` not call the parent to make it impossible to pick up

**Comment by unid15** [comment]
shh

**Comment by icantthinkofanameritenow** [comment]
@Skullyton 
Perhaps it would be helpful to "borrow" some of the stuff Goonstation did with their spacepods.

**Comment by ghost** [comment]
@icantthinkofanameritenow You know the problems you had when porting something from TG? Consider that we are far far far closer to TG than goon. It'd be nigh impossible at this point to truly borrow anything, and would be far easier to just make our own.

**Comment by icantthinkofanameritenow** [comment]
@DrCelt 
Yeah...that brings back some unpleasant memories. 

Still, it's something we can use for inspiration. If nothing else, it does seem kind of odd that there's only one armor type for space pods that can be manufactured. But that's another issue entirely.

**Comment by unid15** [comment]
@ihadtoregisterforthis what do you mean? The structures only spawn in space, and you're free to avoid them if you're not well equipped enough

**Comment by PJB3005** [comment]
I am making a thing!

**Comment by ihadtoregisterforthis** [comment]
@unid15 Many simple mobs can walk on space, and if they start on the station z-level they can eventually reach it with ease.
Other use for these categories would be so some exploration is needed for the best stuff instead of, for example, ordering a borg to RCD some floors in space for loot.

In addition smaller rooms catch fewer meteors than large chunks of derelict arks.

**Comment by unid15** [comment]
I'll probably remove vaults from z-1 then

**Comment by Shadowmech88** [comment]
Or remove the ability of simple mobs to fly through space.

**Comment by PJB3005** [comment]
https://github.com/unid15/vgstation13/pull/5

**Comment by ghost** [comment]
Wait, we can do PRs to this PR?

Oh boy, time for shenanigans

**Comment by MusgardTiger** [comment]
FEISTY

**Comment by NigglyWiggly** [comment]
I have some ideas for these and I might start throwing some stuff together if I can figure out how to into mapping.
- Derelict Space port 'Argo' with some lower tech stuff inside, flashing neon sign outside the port
- Remains of a Grey space craft, (I finally have an excuse to try to put my various Grey themed sprites in the game)
- Maybe an ancient (modern) Satellite? 
- Various space bars, 'gas' stations, and 'truck' stops
- Also as @Jeroen52 said, I'm probably going to just _have_ to make an abandoned space library and arcade

**Comment by Shadowmech88** [comment]
One of the vaults should be a nonfunctional ship that can be restored and piloted.
By "ship" I'm thinking more of the constructable shuttles that can be done via adminbus, rather than the salvage shuttle.

**Comment by unid15** [comment]
Removed all vaults from the station's z-level, so meteor waves wouldn't get affected by them. Stopped ranged hivebots from leaving the hivebot factory (by turning the ones outside the doors into stationary hivebot sentries)
Will squash commits when [NFD]

**Comment by clusterfack** [comment]
Fix conflicts and squash then its good to go

**Comment by unid15** [comment]
Alright all done

(map conflicts = satan)

**Comment by Exxion** [comment]
For the record, we have a .dmm merge engine, so just merging the two conflicting branches locally should fix the conflicts unless an individual tile was changed in both of them
I've never tested that though so it may or may not actually work
It's very old

**Comment by unid15** [comment]
I know but the engine didn't work

**Comment by Exxion** [comment]
Oh
Did it say there were conflicts it couldn't fix or did it just not work?

**Comment by unid15** [comment]
I don't remember but I'm pretty sure it tried fixing them and then gave up

**Comment by Exxion** [comment]
Oh good
It's probably not completely broken, then

**Comment by PJB3005** [comment]
HAPPENING

**Comment by jernaumorat** [comment]
ayyyyy

**Pull request merged**